src/backends/plonky2/primitives/merkletree/mod.rs

```diff
@@ -564,53 +681,166 @@ impl Node {
         }
     }
 
-    // adds the leaf at the tree from the current node (self), without computing any hash
-    pub(crate) fn add_leaf(&mut self, lvl: usize, max_depth: usize, leaf: Leaf) -> TreeResult<()> {
+    /// Applies given Merkle tree op without computing hashes.
+    pub(crate) fn apply_op(
+        &mut self,
+        max_depth: usize,
+        op: MerkleTreeOp,
+        key: RawValue,
+        maybe_value: Option<RawValue>,
+    ) -> TreeResult<()> {
+        let key_path = keypath(max_depth, key)?;
+        // Rule out invalid arguments
+        match (op, maybe_value) {
+            (MerkleTreeOp::Insert, None) | (MerkleTreeOp::Update, None) => {
+                Err(TreeError::invalid_state_transition_proof_arg(format!(
+                    "{:?} op requires a value argument.",
+                    op
+                )))
+            }
+            (MerkleTreeOp::Delete, Some(_)) => {
+                Err(TreeError::invalid_state_transition_proof_arg(format!(
+                    "{:?} op requires no value argument, yet one was provided.",
+                    op
+                )))
+            }
+            _ => Ok(()),
+        }?;
+
+        // Loop through to leaf.
+        self.apply_op_loop(0, max_depth, op, key, &key_path, maybe_value)?;
+
+        // If we are dealing with a deletion, normalise along key
+        // path.
+        match op {
+            MerkleTreeOp::Delete => self.normalise_path(&key_path),
+            _ => Ok(()),
+        }
+    }
+
+    /// Normalises a Merkle tree along a specified path. Useful
+    /// post-deletion.
+    fn normalise_path(&mut self, key_path: &[bool]) -> TreeResult<()> {
```

Comment:
nitpick: this function always returns `Ok(())`, so maybe there's no need to return `TreeResult<()>`?